### PR TITLE
Moving comment api for pipeline instance to spark

### PR DESCRIPTION
**Pull request by bdpiparva**
## Update comment

```bash
curl "http://localhost:8153/go/api/pipelines/spaceship/instance/1/comment" \
	-u 'admin:badger' \
	-H 'Accept: application/vnd.go.cd.v1+json' \
	-H 'Content-Type: application/json' \
	-X POST -d '{ "comment": "Failed bacause of flaky tests"}'

//output
{
  "message" : "Comment successfully updated."
}
```

## Non-integer pipeline counter

```bash
curl "http://localhost:8153/go/api/pipelines/spaceship/instance/one/comment" \
	-u 'admin:badger' \
	-H 'Accept: application/vnd.go.cd.v1+json' \
	-H 'Content-Type: application/json' \
	-X POST -d '{ "comment": "Failed bacause of flaky tests"}'

//output
{"message": "Your request could not be processed. The pipeline counter should be an integer."}
```

## Non-existing pipeline run

```bash
curl "http://localhost:8153/go/api/pipelines/spaceship/instance/9999/comment" \
	-u 'admin:badger' \
	-H 'Accept: application/vnd.go.cd.v1+json' \
	-H 'Content-Type: application/json' \
	-X POST -d '{ "comment": "Failed bacause of flaky tests"}'

//output
{"message":"Pipeline instance for 'spaceship' with counter '9999' were not found!"}
```

##  When the feature is turned off

```bash
curl "http://localhost:8153/go/api/pipelines/spaceship/instance/9999/comment" \
	-u 'admin:badger' \
	-H 'Accept: application/vnd.go.cd.v1+json' \
	-H 'Content-Type: application/json' \
	-X POST -d '{ "comment": "Failed bacause of flaky tests"}'

//output
{"message":"'Pipeline Comment' feature is turned off. Please turn it on to use it."}
```

### CODE REVIEW

File: api/api-pipeline-instance-v1/src/main/java/com/thoughtworks/go/apiv1/pipelineinstance/PipelineInstanceControllerV1.java [hunk 1]
```diff
@@ -68,12 +69,15 @@ public void setupRoutes() {
 
             before(Routes.PipelineInstance.INSTANCE_PATH, mimeType, apiAuthenticationHelper::checkPipelineViewPermissionsAnd403);
             before(Routes.PipelineInstance.HISTORY_PATH, mimeType, apiAuthenticationHelper::checkPipelineGroupOperateOfPipelineOrGroupInURLUserAnd403);
+            before(Routes.PipelineInstance.COMMENT_PATH, mimeType, apiAuthenticationHelper::checkPipelineGroupOperateOfPipelineOrGroupInURLUserAnd403);
```
Comment (kritika-singh3) [inline api/api-pipeline-instance-v1/src/main/java/com/thoughtworks/go/apiv1/pipelineinstance/PipelineInstanceControllerV1.java]: Do we need `operate` permissions for commenting.  Can't `view` be enough for this?
cc:\ @maheshp @adityasood @arvindsv

File: server/src/main/java/com/thoughtworks/go/server/service/PipelineHistoryService.java [hunk 2]
```diff
@@ -553,17 +547,20 @@ private void populateMaterialRevisionsOnBuildCause(PipelineInstanceModel model)
         model.setMaterialRevisionsOnBuildCause(materialRepository.findMaterialRevisionsForPipeline(model.getId()));
     }
 
-    public void updateComment(String pipelineName, int pipelineCounter, String comment, Username username, HttpLocalizedOperationResult result) {
+    public void updateComment(String pipelineName, int pipelineCounter, String comment, Username username) {
```
Comment (kritika-singh3) [inline server/src/main/java/com/thoughtworks/go/server/service/PipelineHistoryService.java]: This method is used in the old comment API. Do we plan to remove that API or else we might have to update it.

Comment (bdpiparva) [reply server/src/main/java/com/thoughtworks/go/server/service/PipelineHistoryService.java]: Yes, I am planning to remove that API. If not then I will update the rails controller

Comment (kritika-singh3) [reply server/src/main/java/com/thoughtworks/go/server/service/PipelineHistoryService.java]: Can we do that in this PR itself, else the build might fail.

File: api/api-pipeline-instance-v1/src/main/java/com/thoughtworks/go/apiv1/pipelineinstance/PipelineInstanceControllerV1.java (hunk 1 above)
Comment (bdpiparva) [reply api/api-pipeline-instance-v1/src/main/java/com/thoughtworks/go/apiv1/pipelineinstance/PipelineInstanceControllerV1.java]: @kritika-singh3 - I have kept the permission to operate as it is the same in the service. I read https://github.com/gocd/gocd/pull/699 but I was not able to figure out why we decided to go with `operate` permission

File: server/src/main/java/com/thoughtworks/go/server/service/PipelineHistoryService.java (hunk 2 above)
Comment (bdpiparva) [reply server/src/main/java/com/thoughtworks/go/server/service/PipelineHistoryService.java]: Done

File: api/api-pipeline-instance-v1/src/main/java/com/thoughtworks/go/apiv1/pipelineinstance/PipelineInstanceControllerV1.java (hunk 1 above)
Comment (maheshp) [reply api/api-pipeline-instance-v1/src/main/java/com/thoughtworks/go/apiv1/pipelineinstance/PipelineInstanceControllerV1.java]: I think we should retain the behavior as is and allow only users with operate permission to comment.

Comment (kritika-singh3) [reply api/api-pipeline-instance-v1/src/main/java/com/thoughtworks/go/apiv1/pipelineinstance/PipelineInstanceControllerV1.java]: ok.

**Pull request merged**